File: Supervised Learning/Ensemble Learning/Bagging&Pasting/Random Forest/coding.py
from sklearn.datasets import load_breast_cancer, load_iris
from sklearn.model_selection import train_test_split
from sklearn.ensemble import (
    BaggingRegressor,
    RandomForestClassifier,
    RandomForestRegressor,
)
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import precision_score, recall_score, r2_score

# Regression

df = load_breast_cancer()

X, y = df.data, df.target

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

random_reg_model = RandomForestRegressor(n_estimators=500, random_state=42)
random_reg_model.fit(X_train, y_train)
y_pred = random_reg_model.predict(X_test)

print("R2 Score: ", r2_score(y_test, y_pred))

for score, name in zip(random_reg_model.feature_importances_, df.feature_names):
    print(round(score, 2), name)

# Or

bagging_reg_model = BaggingRegressor(
    DecisionTreeRegressor(), n_estimators=500, random_state=42
)

bagging_reg_model.fit(X_train, y_train)
y_pred = bagging_reg_model.predict(X_test)

print("R2 Score: ", r2_score(y_test, y_pred))


# classification

df = load_iris()

X, y = df.data, df.target

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

random_class_model = RandomForestClassifier(n_estimators=400, random_state=42)
random_class_model.fit(X_train, y_train)
y_pred = random_class_model.predict(X_test)

print("precision: ", precision_score(y_test, y_pred, average="weighted"))
print("recall: ", recall_score(y_test, y_pred, average="weighted"))

File: Supervised Learning/Ensemble Learning/Bagging&Pasting/Random Forest/coding.R
# Load required libraries
library(randomForest)
library(caret)
library(ipred)
library(datasets)


# Load datasets
data(mtcars)
data(iris)

# Function to split data
split_data <- function(X, y, seed = 42){
  set.seed(seed)
  sample <- sample(1:nrow(X), 0.8 * nrow(X))
  
  X_train <- X[sample, ]
  X_test <- X[-sample, ]
  y_train <- y[sample]
  y_test <- y[-sample]
  
  return(list(X_train = X_train, X_test = X_test, y_train = y_train, y_test = y_test))
}

# Regression

# Prepare data
X <- mtcars[, -1]
y <- mtcars$mpg


# Split data
split <- split_data(X, y)


# Build random forest model
random_reg_model <- randomForest(x = split$X_train, y = split$y_train, ntree = 500)


# Make predictions
y_pred <- predict(random_reg_model, newdata = split$X_test)


# Calculate R-squared
cat("R2 Score: ", round(R2(y_pred, split$y_test), 2))


# Or
bagging_model <- bagging(
  formula = y ~ .,
  data = data.frame(y = split$y_train, split$X_train),
  nbagg = 500
)


# Make predictions with bagging model
y_pred_bagging <- predict(bagging_model, newdata = split$X_test)



# classification


# Prepare data
X <- iris[, -5]
y <- iris$Species


# Split data
split <- split_data(X, y)

# Build random forest model
random_class_model <- randomForest(x = split$X_train, y = split$y_train, ntree = 500)


# Make predictions
y_pred <- predict(random_class_model, newdata = split$X_test)


# Make confusion Matrix
confusion_Matrix <- confusionMatrix(y_pred, split$y_test)
print(confusion_Matrix)

